Jdenticons size and speed test results

elrido opened the pull request:
Sorry, in https://github.com/PrivateBin/PrivateBin/issues/793#issuecomment-856428488 I'd promised to also do the size and speed tests for Jdenticon, which I forgot in that PR.

I have to say that I am disappointed, having now done them and would like to revert making it the new default.

Obviously it is still good to have the extra option, so that folks can switch to it if they like the new look or an issue is discovered in one of the two older libraries.

## Changes
* updated the IconCheck script to include the new library (also for the unused SVG mode - for comparison)
* reverted making Jdenticon the default, sticking with Identicon for now

TL;DR: Jdenticon takes between 9 (PHP 8.1) and 14 (PHP 7.4) times as long to generate the PNGs over Identicon and they are about double the size, but still a little smaller then vizhash's.

Results for PHP 7.4:
```
$ docker run -ti -v ~/git/PrivateBin:/srv:ro alpine:3.13
# apk add php7 php7-gd php7-opcache php7-pecl-imagick
# cd /srv
# tst/IconTest
generate 100000 hmacs and pre-populate the result array, so tests wont be slowed down
run vizhash tests
run identicon GD tests
run identicon ImageMagick tests
run identicon SVG tests
run jdenticon GD tests
run jdenticon SVG tests

Generator:            	 min	 max	 avg	seconds
──────────────────────	────	────	────	───────
vizhash:              	 162	 734	 457	 13.588
identicon GD:         	 170	 270	 225	  8.639
identicon ImageMagick:	 222	 310	 283	385.331
identicon SVG:        	 370	 802	 643	  1.995
jdenticon GD:         	 286	 518	 418	120.633
jdenticon SVG:        	 934	1762	1261	 13.414
# php --version
PHP 7.4.26 (cli) (built: Nov 18 2021 21:41:28) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.26, Copyright (c), by Zend Technologies
```

Results for PHP 8.1:
(I wanted to test it on PHP 8.2, but alpine doesn't yet provide a php82-pecl-imagick package)
```
$ docker run -ti -v ~/git/PrivateBin:/srv:ro alpine:edge
# apk add php81 php81-gd php81-opcache php81-pecl-imagick
# cd /srv
# tst/IconTest
generate 100000 hmacs and pre-populate the result array, so tests wont be slowed down
run vizhash tests
run identicon GD tests
run identicon ImageMagick tests
run identicon SVG tests
run jdenticon GD tests
run jdenticon SVG tests

Generator:            	 min	 max	 avg	seconds
──────────────────────	────	────	────	───────
vizhash:              	 162	 742	 456	 18.225
identicon GD:         	 170	 266	 225	  8.176
identicon ImageMagick:	 222	 310	 283	460.428
identicon SVG:        	 370	 802	 643	  1.332
jdenticon GD:         	 286	 518	 417	 74.391
jdenticon SVG:        	 934	1762	1260	  9.774
# php --version
PHP 8.1.12 (cli) (built: Oct 25 2022 19:18:03) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.12, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.12, Copyright (c), by Zend Technologies
```

Specs for reference:
```
$ lscpu | grep name
Modelname:                       AMD Ryzen 5 3400G with Radeon Vega Graphics
```

For reference, here are results from a different system and PHP 7.2, from when we considered switching to SVGs:
https://github.com/PrivateBin/PrivateBin/issues/148#issuecomment-502426961

r4sas commented:
Can you share test script with jdenticon support?
Also I can't understand why you write `GD` as renderer, if there is no GD support (internal generator, Imagick and SVG only).

elrido commented:
> Can you share test script with jdenticon support?

It is included in the PR and found in https://github.com/PrivateBin/PrivateBin/blob/b2ef205411984f2c40a075da0774123ddfa61335/tst/IconTest

> Also I can't understand why you write `GD` as generator, if there is no GD support (internal generator, Imagick and SVG only).

You are right, they don't use GD, it's a [pure PHP implementation](https://github.com/dmester/jdenticon-php/blob/master/src/Canvas/Png/PngEncoder.php) of the [PNG format](https://github.com/dmester/jdenticon-php/blob/master/src/Canvas/Png/PngBuffer.php). They offer [optional ImageMagick](https://jdenticon.com/php-api/T_Identicon.html#enableimagemagick) support and will autodetect and enable it in PHP 5, but don't recommend it for PHP 7:

> > The internal renderer has no external dependencies, and is recommended for PHP 7 and later. For PHP 5, ImageMagick is recommended for better performance.

So I've updated the script to have clearer labels and test all 3 possibilities and on the same machine as for the above tests I get:

```
# same docker install commands as above
generate 100000 hmacs and pre-populate the result array, so tests wont be slowed down
run vizhash tests
run identicon GD tests
run identicon ImageMagick tests
run identicon SVG tests
run jdenticon tests
run jdenticon ImageMagick tests
run jdenticon SVG tests

Generator:            	 min	 max	 avg	seconds
──────────────────────	────	────	────	───────
vizhash:              	 162	 742	 456	 17.671
identicon GD:         	 170	 266	 225	  8.126
identicon ImageMagick:	 222	 310	 283	443.528
identicon SVG:        	 370	 802	 643	  1.378
jdenticon:            	 286	 518	 417	  74.29
jdenticon ImageMagick:	 362	 550	 472	3373.792
jdenticon SVG:        	 934	1762	1261	  9.512
/srv # php --version
PHP 8.1.12 (cli) (built: Oct 25 2022 19:18:03) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.12, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.12, Copyright (c), by Zend Technologies
```

Don't get me wrong, I like having options, these look nice and now we have an option for folks without GD (I've updated the INSTALL.md accordingly). So we could accept the performance hit and leave the new library as the default, allowing folks to use PrivateBin even without GD library.

The images are generated only once per comment creation (which the traffic limiter should rate limit to one per 10 sec per source IP) and then stored. So there is also a little bit of extra storage required, but they are still smaller than the original vizhash images, so that shouldn't matter much. And in any case, they can be turned off.

rugk commented:
Also reported the actual performance issue upstream, as without a report, they have no chance to fix that: https://github.com/dmester/jdenticon-php/issues/8
